File: Day-11/python/rainbowdashlabs/day_11.py
from dataclasses import dataclass

r_monkeys = [e for e in open("input.txt").read().split("\n\n")]


@dataclass(repr=True, unsafe_hash=True)
class Monkey:
    id: int
    items: list[int]
    operation: str
    test: int
    t_true: int
    t_false: int
    inspects: int = 0

    @staticmethod
    def parse(monkey: str) -> "Monkey":
        monkey = monkey.splitlines()
        id = int(monkey[0][-2])
        items = list(map(int, monkey[1].split(":")[1].split(",")))
        operation = monkey[2].split("=")[1].strip()
        test = int(monkey[3].split("by")[1])
        t_true = int(monkey[4][-1])
        t_false = int(monkey[5][-1])
        return Monkey(id, items, operation, test, t_true, t_false)

    def inspect(self, mod=3) -> list[(int, int)]:
        next = []
        for item in self.items:
            new = int(eval(self.operation.replace("old", str(item))))
            new %= mod
            next.append((self.t_true if not new % self.test else self.t_false, new))
            self.inspects += 1
        self.items = []
        return next


monkeys = {monkey.id: monkey for monkey in [Monkey.parse(e) for e in r_monkeys]}

for _ in range(20):
    for monkey in monkeys.values():
        for t, item in monkey.inspect():
            monkeys[t].items.append(item)

monkey_list = list(monkeys.values())
monkey_list.sort(key=lambda x: x.inspects)
print(f"Part 1: {monkey_list[-1].inspects * monkey_list[-2].inspects}")

monkeys = {monkey.id: monkey for monkey in [Monkey.parse(e) for e in r_monkeys]}

mod = 1
for t in monkeys.values():
    mod *= t.test

for _ in range(10000):
    for monkey in monkeys.values():
        for t, item in monkey.inspect(mod):
            monkeys[t].items.append(item)

monkey_list = list(monkeys.values())
monkey_list.sort(key=lambda x: x.inspects)
print(f"Part 2: {monkey_list[-1].inspects * monkey_list[-2].inspects}")

File: Day-11/clojure/JohnnyJayJay/day11.clj
(ns day11
  (:require [clojure.string :as str]))

(defn parse-monkey [[_header items operation & other]]
  (let [items  (map parse-long (re-seq #"\d+" (apply str items other)))
        [items [test then else]] (split-at (- (count items) 3) items)
        [a op b] (->> (re-matches #".+new = (old|\d+) ([*+]) (old|\d+)" operation)
                        (drop 1)
                        (mapv read-string))]
    {:item-queue (vec items)
     :test test
     :then then
     :else else
     :operator (eval op)
     :operands [a b]
     :inspections 0}))

(defn inspect-1 [old {:keys [operator operands]} _test]
  (quot (apply operator (map #(if (symbol? %) old %) operands)) 3))

(defn inspect-2 [old {:keys [operator operands]} test]
  (rem (apply operator (map #(if (symbol? %) old (rem % test)) operands)) test))

(defn parse-input [input]
  (let [monkeys (->> (str/split input #"\n\n")
                     (map str/split-lines)
                     (map parse-monkey))
        items (mapcat :item-queue monkeys)]
    (loop [item-idxs (range (count items))
           [{:keys [item-queue] :as monkey} & remaining] monkeys
           result []]
      (if monkey
        (let [[start-items other-items] (split-at (count item-queue) item-idxs)]
          (recur
           other-items
           remaining
           (conj result (assoc monkey :item-queue (vec start-items) :items (vec items)))))
        result))))

(defn prepare-part-2 [monkeys]
  (mapv
   (fn [{:keys [test] :as monkey}]
     (update monkey :items (partial mapv #(rem % test))))
   monkeys))

(defn monkey-item-turn [inspect-fn idx monkeys item]
  (let [{:keys [test then else] :as monkey} (get monkeys idx)
        updated-monkeys (mapv #(update-in % [:items item] inspect-fn monkey (:test %)) monkeys)
        item-remainder (rem (get-in updated-monkeys [idx :items item]) test)
        new-owner (if (zero? item-remainder) then else)]
    (update-in updated-monkeys [new-owner :item-queue] conj item)))

(defn monkey-turn [inspect-fn monkeys idx]
  (let [{:keys [item-queue]} (get monkeys idx)]
    (-> (reduce
         (partial monkey-item-turn inspect-fn idx)
         monkeys
         item-queue)
        (update-in [idx :inspections] + (count item-queue))
        (assoc-in [idx :item-queue] []))))

(defn simulate-monkeys [inspect-fn monkeys]
  (iterate (fn [monkeys] (reduce (partial monkey-turn inspect-fn) monkeys (range (count monkeys)))) monkeys))

(defn monkey-business [inspect-fn rounds monkeys]
  (->> monkeys
       (simulate-monkeys inspect-fn)
       (drop rounds)
       first
       (map :inspections)
       (sort >)
       (take 2)
       (apply *)))

(defn -main [file & _]
  (let [monkeys (-> file slurp parse-input)]
    (println "Part 1:" (monkey-business inspect-1 20 monkeys))
    (println "Part 2:" (monkey-business inspect-2 10000 (prepare-part-2 monkeys)))))
